File: app/src/main/java/com/dhsdevelopments/potato/clientapi/PotatoApi.java
package com.dhsdevelopments.potato.clientapi;

import com.dhsdevelopments.potato.clientapi.channel2.ChannelsResult;
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationRequest;
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationResult;
import com.dhsdevelopments.potato.clientapi.message.MessageHistoryResult;
import com.dhsdevelopments.potato.clientapi.notifications.PotatoNotificationResult;
import com.dhsdevelopments.potato.clientapi.sendmessage.SendMessageRequest;
import com.dhsdevelopments.potato.clientapi.sendmessage.SendMessageResult;
import com.dhsdevelopments.potato.clientapi.unreadnotification.UpdateUnreadNotificationRequest;
import com.dhsdevelopments.potato.clientapi.unreadnotification.UpdateUnreadNotificationResult;
import com.dhsdevelopments.potato.clientapi.users.LoadUsersResult;
import retrofit.Call;
import retrofit.http.*;

import java.util.Map;

public interface PotatoApi
{
    @GET( "channels2" )
    Call<ChannelsResult> getChannels2( @Header( "API-token" ) String apiKey );

    @GET( "channel/{cid}/history?format=json" )
    Call<MessageHistoryResult> loadHistoryAsJson( @Header( "API-token" ) String apiKey,
                                                  @Path( "cid" ) String channelId,
                                                  @Query( "num" ) int numMessages );

    @GET( "channel-updates?format=json" )
    Call<PotatoNotificationResult> channelUpdates( @Header( "API-token" ) String apiKey,
                                                   @Query( "channels" ) String channels,
                                                   @Query( "services" ) String services,
                                                   @Query( "event-id" ) String eventId );

    @POST( "channel-updates/update" )
    Call<ChannelUpdatesUpdateResult> channelUpdatesUpdate( @Header( "API-token" ) String apiKey,
                                                           @Query( "event-id" ) String eventId,
                                                           @Query( "cmd" ) String cmd,
                                                           @Query( "channel" ) String channelId,
                                                           @Query( "services" ) String services );

    @POST( "channel/{cid}/create" )
    Call<SendMessageResult> sendMessage( @Header( "API-token" ) String apiKey,
                                         @Path( "cid" ) String channelId,
                                         @Body SendMessageRequest request );

//    @Multipart
//    @POST( "channel/{cid}/upload" )
//    Call<SendMessageResult> sendMessageWithFile( @Header( "API-token" ) String apiKey,
//                                                 @Path( "cid" ) String channelId,
//                                                 @Part("content") SendMessageRequest request,
//                                                 @Part("body") RequestBody fileBody );

    @Multipart
    @POST( "channel/{cid}/upload" )
    Call<SendMessageResult> sendMessageWithFile( @Header( "API-token" ) String apiKey,
                                                 @Path( "cid" ) String channelId,
                                                 @PartMap Map<String, Object> params );

    @GET( "channel/{cid}/users" )
    Call<LoadUsersResult> loadUsers( @Header( "API-token" ) String apiKey,
                                     @Path( "cid" ) String channelId );

    @POST( "register-gcm" )
    Call<GcmRegistrationResult> registerGcm( @Header( "API-token" ) String apiKey,
                                             @Body GcmRegistrationRequest request );

    @POST( "channel/{cid}/clear-notifications" )
    Call<ClearNotificationsResult> clearNotificationsForChannel( @Header( "API-token" ) String apiKey,
                                                                 @Path( "cid" ) String channelId );

    @POST( "channel/{cid}/unread-notification" )
    Call<UpdateUnreadNotificationResult> updateUnreadNotification( @Header( "API-token" ) String apiKey,
                                                                   @Path( "cid" ) String channelId,
                                                                   @Body UpdateUnreadNotificationRequest request );
}

File: app/src/main/java/com/dhsdevelopments/potato/clientapi/PotatoApi.kt
package com.dhsdevelopments.potato.clientapi

import com.dhsdevelopments.potato.clientapi.channel2.ChannelsResult
import com.dhsdevelopments.potato.clientapi.deletemessage.DeleteMessageResult
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationRequest
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationResult
import com.dhsdevelopments.potato.clientapi.message.MessageHistoryResult
import com.dhsdevelopments.potato.clientapi.notifications.PotatoNotificationResult
import com.dhsdevelopments.potato.clientapi.search.SearchResult
import com.dhsdevelopments.potato.clientapi.sendmessage.SendMessageRequest
import com.dhsdevelopments.potato.clientapi.sendmessage.SendMessageResult
import com.dhsdevelopments.potato.clientapi.unreadnotification.UpdateUnreadNotificationRequest
import com.dhsdevelopments.potato.clientapi.unreadnotification.UpdateUnreadNotificationResult
import com.dhsdevelopments.potato.clientapi.users.LoadUsersResult
import retrofit.Call
import retrofit.http.*

interface PotatoApi {
    @GET("channels2")
    fun getChannels2(@Header("API-token") apiKey: String): Call<ChannelsResult>

    @GET("channel/{cid}/history?format=json")
    fun loadHistoryAsJson(@Header("API-token") apiKey: String,
                          @Path("cid") channelId: String,
                          @Query("num") numMessages: Int,
                          @Query("from") from: String): Call<MessageHistoryResult>

    @GET("channel-updates?format=json")
    fun channelUpdates(@Header("API-token") apiKey: String,
                       @Query("channels") channels: String,
                       @Query("services") services: String,
                       @Query("event-id") eventId: String?): Call<PotatoNotificationResult>

    @POST("channel-updates/update")
    fun channelUpdatesUpdate(@Header("API-token") apiKey: String,
                             @Query("event-id") eventId: String,
                             @Query("cmd") cmd: String,
                             @Query("channel") channelId: String,
                             @Query("services") services: String): Call<ChannelUpdatesUpdateResult>

    @POST("channel/{cid}/create")
    fun sendMessage(@Header("API-token") apiKey: String,
                    @Path("cid") channelId: String,
                    @Body request: SendMessageRequest): Call<SendMessageResult>

    @DELETE("message/{messageId}")
    fun deleteMessage(@Header("API-token") apiKey: String,
                      @Path("messageId") messageId: String): Call<DeleteMessageResult>

    @Multipart
    @POST("channel/{cid}/upload")
    fun sendMessageWithFile(@Header("API-token") apiKey: String,
                            @Path("cid") channelId: String,
                            @PartMap params: Map<String, Any>): Call<SendMessageResult>

    @GET("channel/{cid}/users")
    fun loadUsers(@Header("API-token") apiKey: String,
                  @Path("cid") channelId: String): Call<LoadUsersResult>

    @POST("register-gcm")
    fun registerGcm(@Header("API-token") apiKey: String,
                    @Body request: GcmRegistrationRequest): Call<GcmRegistrationResult>

    @POST("channel/{cid}/clear-notifications")
    fun clearNotificationsForChannel(@Header("API-token") apiKey: String,
                                     @Path("cid") channelId: String): Call<ClearNotificationsResult>

    @POST("channel/{cid}/unread-notification")
    fun updateUnreadNotification(@Header("API-token") apiKey: String,
                                 @Path("cid") channelId: String,
                                 @Body request: UpdateUnreadNotificationRequest): Call<UpdateUnreadNotificationResult>

    @GET("channel/{cid}/search")
    fun searchMessages(@Header("API-token") apiKey: String,
                       @Path("cid") channelId: String,
                       @Query("query") query: String,
                       @Query("star-only") starOnly: String): Call<SearchResult>
}
